### ISSUE TITLE
Incorrect transaction id reporting when using multiple threads

### ISSUE THREAD

**Issue by achamayou** [microsoft/CCF#1163]
https://dev.azure.com/MSRC-CCF/CCF/_build/results?buildId=8222&view=ms.vss-test-web.build-test-results-tab&runId=42682&resultId=100010&paneView=attachments

```
2020-05-08 14:19:34.546 | INFO     | infra.remote:log_errors:74 - 2020-05-08T13:19:29.576005Z        100 [info ] amples/perf_client/perf_client.h:530 | Triggered signature at 2.20
2020-05-08 14:19:34.547 | INFO     | infra.remote:log_errors:74 - 2020-05-08T13:19:29.576026Z        100 [info ] ../samples/perf_client/timing.h:261  | Waiting for transaction ID 2.20
2020-05-08 14:19:34.547 | INFO     | infra.remote:log_errors:74 - 2020-05-08T13:19:30.111115Z        100 [info ] ../samples/perf_client/timing.h:299  | Found global commit 2.20
2020-05-08 14:19:34.547 | INFO     | infra.remote:log_errors:74 - 2020-05-08T13:19:30.111134Z        100 [info ] ../samples/perf_client/timing.h:304  |  (headers term: 2, local: 21, global: 21)
2020-05-08 14:19:34.547 | INFO     | infra.remote:log_errors:74 - 2020-05-08T13:19:32.023628Z        100 [info ] amples/perf_client/perf_client.h:439 | About to begin timing
2020-05-08 14:19:34.547 | INFO     | infra.remote:log_errors:74 - 2020-05-08T13:19:32.023642Z        100 [info ] amples/perf_client/perf_client.h:441 | Began timing
2020-05-08 14:19:34.547 | INFO     | infra.remote:log_errors:74 - 2020-05-08T13:19:33.231753Z        100 [fail ] amples/perf_client/perf_client.h:682 | Transaction exception: Term went backwards (expected 2, saw 0)!
2020-05-08 14:19:34.547 | ERROR    | infra.remote:log_errors:83 - Contents of /mnt/build/1/s/build/workspace/SB_sig_SGX_CFT^_client_0/err:
terminate called after throwing an instance of 'std::exception'
  what():  std::exception
```

**Comment by achamayou** [comment]
Another instance in https://dev.azure.com/MSRC-CCF/CCF/_build/results?buildId=8301&view=ms.vss-test-web.build-test-results-tab&runId=42962&resultId=100014&paneView=attachments

**Comment by eddyashton** [comment]
This only happens when multiple threads are enabled, and under load (ie - in the performance tests, not in any other tests). In `frontend.h`, when populating the headers, we ask for the term of an index _later_ than `last_idx` in Raft, so it believes this is in the future and returns 0. I've added some logging but don't fully understand the behaviour, it seems to be some issue with the reserved index used by signature transactions:

```
2020-05-13T12:56:06.101445Z -1.093 1   [info ] ../src/consensus/raft/raft.h:365     | Setting last_idx to 39722 (replicate)
2020-05-13T12:56:06.101754Z -1.094 1   [info ] ../src/consensus/raft/raft.h:365     | Setting last_idx to 39723 (replicate)
2020-05-13T12:56:06.101920Z -1.084 0   [info ] ../src/node/rpc/../history.h:530     | Signed at 39724 view: 2 commit: 37501
2020-05-13T12:56:06.130564Z -1.112 0   [info ] ../src/consensus/raft/raft.h:365     | Setting last_idx to 39724 (replicate)
2020-05-13T12:56:06.189979Z -1.172 1   [info ] ../src/consensus/raft/raft.h:475     | Asked for 39725, but last is 39724
2020-05-13T12:56:06.190025Z -1.172 1   [fail ] ../src/node/rpc/frontend.h:389       | Uh oh: I'm responding about commit 39725 and I don't know what view it is in!
2020-05-13T12:56:06.190040Z -1.172 1   [fail ] ../src/node/rpc/frontend.h:391       | Got commit from commit version for this 'SmallBank_transact_savings' RPC
2020-05-13T12:56:06.190470Z -1.172 1   [info ] ../src/consensus/raft/raft.h:475     | Asked for 39725, but last is 39724
2020-05-13T12:56:06.190487Z -1.172 1   [fail ] ../src/node/rpc/frontend.h:389       | Uh oh: I'm responding about commit 39725 and I don't know what view it is in!
2020-05-13T12:56:06.190500Z -1.172 1   [fail ] ../src/node/rpc/frontend.h:391       | Got commit from read version for this 'SmallBank_balance' RPC
2020-05-13T12:56:06.190965Z -1.173 1   [info ] ../src/consensus/raft/raft.h:365     | Setting last_idx to 39725 (replicate)
2020-05-13T12:56:06.191013Z -1.173 1   [info ] ../src/consensus/raft/raft.h:365     | Setting last_idx to 39726 (replicate)
```

**Comment by eddyashton** [comment]
In `kv::Store::commit` we ensure that only consecutive transactions are replicated. When a signature has booked a version but not yet committed, an executing transaction may get a later version and be put into the `pending_txs` queue. This transaction will return `CommitSuccess::OK` before it is replicated, so before we can retrieve its term from the consensus (it will replicate as part of a later transaction). I see a few potential solutions, but none sounds quite right:

- If consensus doesn't know the term for a transaction we just executed, spin for a while until it does.
- Assume these missing terms must have happened in the current term.

Both approaches sound like they will return the wrong result if we have an election in the interim. It seems that it is possible to execute a transaction reading the state `2.5`, but if the thread is stalled and `2.5` is replaced, we'd eventually call this `3.5`? I'll give this some more though.

Separately but perhaps a contributing factor we're regularly producing multiple signatures at near identical points. Tracked in #1176

**Comment by eddyashton** [comment]
More logs to show the problematic flow.

A normal, working transaction takes a version, builds a batch of just that version (the next is not available yet), and replicated:
```
2   [info ] ../src/kv/kv.h:1079                  | Normal tx claimed version 8728
2   [info ] ../src/kv/kv.h:1828                  | Reached first missing version 8729
2   [info ] ../src/kv/kv.h:1863                  | Replicating batch of 1 transactions
2   [info ] ../src/consensus/raft/raft.h:365     | Setting last_idx to 8728 (replicate)
```

When a signature leaves a hole:
```
0   [info ] ../src/node/rpc/../history.h:530     | Signed at 8729 view: 2 commit: 7501
2   [info ] ../src/kv/kv.h:1079                  | Normal tx claimed version 8730
0   [info ] ../src/kv/kv.h:1828                  | Reached first missing version 8730
0   [info ] ../src/kv/kv.h:1863                  | Replicating batch of 1 transactions
0   [info ] ../src/consensus/raft/raft.h:365     | Setting last_idx to 8729 (replicate)
2   [info ] ../src/kv/kv.h:1828                  | Reached first missing version 8729
2   [info ] ../src/kv/kv.h:1858                  | Empty replication batch
2   [info ] ../src/consensus/raft/raft.h:475     | Asked for 8730, but last is 8729
2   [fail ] ../src/node/rpc/frontend.h:389       | Uh oh: I'm responding about commit 8730 and I don't know what view it is in!
2   [fail ] ../src/node/rpc/frontend.h:391       | Got commit from commit version for this 'SmallBank_write_check' RPC
2   [info ] ../src/kv/kv.h:1079                  | Normal tx claimed version 8731
2   [info ] ../src/kv/kv.h:1828                  | Reached first missing version 8732
2   [info ] ../src/kv/kv.h:1863                  | Replicating batch of 2 transactions
2   [info ] ../src/consensus/raft/raft.h:365     | Setting last_idx to 8730 (replicate)
2   [info ] ../src/consensus/raft/raft.h:365     | Setting last_idx to 8731 (replicate)
```

**Comment by eddyashton** [comment]
> * Assume these missing terms must have happened in the current term.

This looked like a promising approach. Raft can report that all future unknown commits happen in the current term, and then the callers above them can check whether the commit actually happened, if necessary.

The problem with this is that there is still a gap between a Tx reporting commit success (either because it has been replicated, or because it has been queued as pending), and asking for its term. It is technically possible for a full election to occur in this time, in which case the executed transaction which was _really_ 2.5 has been rolled back, but we could tell the caller it was 3.5.

I believe the only way around this is an earlier binding between term and version. It should be possible to ask consensus for its current term _before_ replicating. Then the Tx knows it _would_ be 2.5 _if it were replicated_, it is safe to report this to users. The consensus can discard any replication requests in earlier terms (though this isn't an issue since they would be in batches are already dropped by the kv during rollback). Getting the current term without deadlock is potentially difficult.

**Comment by achamayou** [comment]
I think this is closed with #1252. There are still some test issues associated with threaded builds, but nothing that looks related to this.
